File: 18-4sum/18-4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) { 
    nums=nums.sort((a,b)=>a-b);
    let result=[]
    for(let i=0;i<nums.length-3;i++){
        if(nums[i]===nums[i-1]) continue;
        for(let j=i+1;j<nums.length-2;j++){
            if(nums[j]===nums[j-1]&&j>i+1) continue;
            let left=j+1;
            let right=nums.length-1;
            while(left<right){
                const sum=nums[i]+nums[j]+nums[left]+nums[right];
                if(sum===target){
                    result.push([nums[i],nums[j],nums[left],nums[right]]);
                    while(nums[right]===nums[right-1]&&right>left) right--;
                    while(nums[left]===nums[left+1]&&right>left) left++;
                    right--;
                    left++;
                    continue;
                }
                if(sum>=target){
                    while(nums[right]===nums[right-1]&&right>left) right--;
                    right--;
                    continue;
                }else{
                    while(nums[left]===nums[left+1]&&right>left) left++;
                    left++;
                    continue;
                }
        }
        }
       
        
      
        
    }
    return result;
};

File: 18-4sum/18-4sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """  
        nums.sort()
        result=[]
        for i in range(len(nums)-2):
            # if nums[i]==nums[i-1]: continue
            for j in range(i+1,len(nums)-1):
                left=j+1
                right=len(nums)-1
                while left<right:
                    sum=nums[i]+nums[j]+nums[left]+nums[right]
                    arr=[nums[i],nums[j],nums[left],nums[right]]
                    if sum==target:
                        if arr not in result:
                            result.append(arr)
                        left+=1
                        right-=1

                    elif sum<target:
                        left+=1
                    else:
                        right-=1
                        
        
        return result
            
                
              
      
        